### PR TITLE
add cloning

### PR DESCRIPTION
**Pull request by shellderp**
A lot is copy paste from the create form/endpoint

image

image

### CODE REVIEW

File: service/src/main/kotlin/app/cash/backfila/dashboard/CloneBackfillAction.kt [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+      throw BadRequestException("batch_size must be >= 1")
+    }
+    if (request.scan_size < request.batch_size) {
+      throw BadRequestException("scan_size must be >= batch_size")
+    }
+    if (request.extra_sleep_ms < 0) {
+      throw BadRequestException("extra_sleep_ms must be >= 0")
+    }
+    request.backoff_schedule?.let { schedule ->
+      if (schedule.split(',').any { it.toLongOrNull() == null }) {
```
Comment (mpawliszyn) [inline service/src/main/kotlin/app/cash/backfila/dashboard/CloneBackfillAction.kt:95]: Perhaps there should be a more general place for these validations?

File: service/src/main/kotlin/app/cash/backfila/dashboard/CloneBackfillAction.kt [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+          )
+          session.save(dbRunPartition)
+        }
+      } else {
+        // Verify partitions match source backfill, use per partition ranges.
+        val sourceBackfill = session.load<DbBackfillRun>(Id(id))
+        // Source partitions have to match new partitions.
+        val sourcePartitions = sourceBackfill.partitions(session, queryFactory)
+        if (partitions.map { it.partition_name }.toSet() != sourcePartitions.map { it.partition_name }.toSet()) {
+          throw BadRequestException("Can't clone backfill ranges from `$id`, newly computed partitions don't match.")
```
Comment (mpawliszyn) [inline service/src/main/kotlin/app/cash/backfila/dashboard/CloneBackfillAction.kt:180]: Maybe mention if you want to clone to create a NEW one?

File: service/src/main/kotlin/app/cash/backfila/dashboard/CloneBackfillAction.kt [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+          val dbRunPartition = DbRunPartition(
+              backfillRun.id,
+              sourcePartition.partition_name,
+              sourcePartition.backfillRange(),
+              backfillRun.state,
+              sourcePartition.estimated_record_count
+          )
+          // Copy the cursor if continuing, otherwise just leave blank to start from beginning.
+          if (request.range_clone_type == RangeCloneType.CONTINUE) {
+            dbRunPartition.pkey_cursor = sourcePartition.pkey_cursor
```
Comment (mpawliszyn) [inline service/src/main/kotlin/app/cash/backfila/dashboard/CloneBackfillAction.kt:193]: If its null does it still use the start range? There should probably be a test for this to prove this.

Comment (shellderp) [reply service/src/main/kotlin/app/cash/backfila/dashboard/CloneBackfillAction.kt:193]: null = start from the beginning

when null cursor is passed, the operator figures out the cursor from the range start

File: service/src/main/kotlin/app/cash/backfila/dashboard/CloneBackfillAction.kt (hunk 1 above)
Comment (shellderp) [reply service/src/main/kotlin/app/cash/backfila/dashboard/CloneBackfillAction.kt:95]: in a followup I can probably refactor to have some common code between create/clone

File: service/src/main/kotlin/app/cash/backfila/dashboard/CloneBackfillAction.kt (hunk 2 above)
Comment (shellderp) [reply service/src/main/kotlin/app/cash/backfila/dashboard/CloneBackfillAction.kt]: good idea, done

File: service/web/tabs/app/src/containers/BackfillStatusContainer.tsx [hunk 4]
```diff
@@ -84,6 +93,10 @@ class BackfillStatusContainer extends React.Component<
     this.setState({ editing: {} })
   }
 
+  componentWillUnmount() {
```
Comment (shellderp) [inline service/web/tabs/app/src/containers/BackfillStatusContainer.tsx:96]: btw.. fixed this. Previously it kept polling status even after unloading the page. oops!

File: service/web/tabs/app/src/containers/CloneFormContainer.tsx [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+      batch_size: 100,
+      num_threads: 1,
+      range_clone_type: "RESTART",
+      pkey_range_start: null,
+      pkey_range_end: null,
+      backoff_schedule: null,
+      extra_sleep_ms: 0,
+      parameters: {}
+    })
+    Axios.get(`/backfills/${this.id}/status`)
```
Comment (shellderp) [inline service/web/tabs/app/src/containers/CloneFormContainer.tsx:79]: axios is easy, I'll probably replace all the http with this

File: service/src/main/kotlin/app/cash/backfila/dashboard/CloneBackfillAction.kt (hunk 1 above)
Comment (mpawliszyn) [reply service/src/main/kotlin/app/cash/backfila/dashboard/CloneBackfillAction.kt:95]: sounds good.

File: service/web/tabs/app/src/containers/BackfillStatusContainer.tsx (hunk 4 above)
Comment (mpawliszyn) [reply service/web/tabs/app/src/containers/BackfillStatusContainer.tsx:96]: nice!

File: service/web/tabs/app/src/containers/CloneFormContainer.tsx (hunk 5 above)
Comment (mpawliszyn) [reply service/web/tabs/app/src/containers/CloneFormContainer.tsx:79]: I'll have to look this up.

**Pull request merged**